lib/internal/modules/cjs/loader.js

```diff
@@ -651,14 +651,7 @@ Module.prototype._compile = function(content, filename) {
 
   content = stripShebang(content);
 
-  // create wrapper function
-  var wrapper = Module.wrap(content);
-
-  var compiledWrapper = vm.runInThisContext(wrapper, {
-    filename: filename,
-    lineOffset: 0,
-    displayErrors: true
-  });
+  const compiledWrapper = vm.getWrappedFunction(content, filename);
```

Comment:
Would there be a way to have the cake and eat it to? As is the removal of `Module.wrap` use is a no-go for me since CJS tools rely on it. If there is a way to get the benefit + allow `Module.wrap` tie-ins that would be rad.

This could be done in a variety of ways. For example using the `Module.wrapper[0]` and `Module.wrapper[1]` to be used to compile something internally (waves hands as I'm not familiar with V8 internals). Or having setters on `Module.wrapper`, `Module.wrapper[0]`, and `Module.wrapper[1]` to trigger the old non-`vm.getWrappedFunction` use.

Reply:
This is PR addresses a minor inconvenience that has existed since Node's creation so I'm not OK dropping any existing CJS tie-ins here. `Module.wrap` and `Module.wrapper` have been used by tooling to customize CJS module construction so I'd be wary of removing them. CJS is essentially frozen so I'd urge you to reconsider.

Reply:
@jdalton Proper syntax error messages that don’t get obfuscated by the added function code are important as well.

If there are specific use cases you’re thinking of, it would be good to know what public API would address those (and possibly whether they can also be addressed by monkey-patching the newly introduced `vm` method)

Reply:
@jdalton what specifically makes `Module.wrap` special to an extent that CJS tools "rely" on it? I'd love to see code that works with `Module.wrap` but breaks on what we have right now. Aiming for a `semver-minor` here, I'll be doing whatever I can to make sure no userland code breaks (and technically, it shouldn't, even right now).

I mean, you very well know at this point about all problems that the original function causes and how this will be an improvement, but again, I'd be making sure that nothing breaks in the process and if something does, let me know.

Reply:
> Or having setters on Module.wrapper, Module.wrapper[0], and Module.wrapper[1] to trigger the old non-vm.getWrappedFunction use.

Reply:
@ryzokuken

As @devsnek points out I've given an example of a couple ways to maintain CJS compatibility and still benefit from the newness. I'd appreciate it if a little more effort were put into preserving existing, CJS bits is all. Again, the issue solved by this PR has existed since CJS-was a thing, so while it would be nice I'd like to not lose anything in the realm of CJS while tackling it. I think it's possible to solve the issue while still keeping the CJS tie-ins relatively unaffected. So I'm rooting for that.

Reply:
just to clarify i'm in favour of removing the wrapper.

Reply:
@jdalton again, I'm removing neither `Module.wrap` nor `Module.wrapper` here.

If you could point out which CJS modules/tools broke while changing `Module._compile` internally, I could make sure it wouldn't.

Reply:
I think what John is referring to isn’t that `Module.wrap` or `Module.wrapper` don’t work anymore, it’s that *replacing*/monkey-patching them doesn’t have the effect of actually changing Node’s behaviour anymore?

Reply:
Wait, CJS modules are monkey-patching the `Module` object and we're holding up PRs because it won't allow them to monkey-patch anymore!?

Reply:
> Wait, CJS modules are monkey-patching the Module object and we're holding up PRs because it won't allow them to monkey-patch a core module anymore!?

Yes, it’s the sensible thing to do for us. We have Node’s past lack of proper hook APIs to blame for this, not users who need to get things to work…

Reply:
CJS is very monkey patchable. This particular API and functionality has existed since Jan 2011 _(so pretty old)_. Losing the ability to monkey patch in this way is why I'm raising the concern.